tests/events/main.c

```diff
@@ -0,0 +1,80 @@
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#define NUM_EVENTS 4
+#define INTERRUPT_AT 1
+
+struct info {
+	int n;
+};
+
+static int
+event(void * cookie)
+{
+	struct info * I = cookie;
+
+	printf("event %u\n", I->n);
+
+	/* Quit via interrupting. */
+	if (I->n == INTERRUPT_AT)
+		raise(SIGUSR1);
+
+	/* Success! */
+	return (0);
+}
+
+int
+main(int argc, char * argv[])
+{
+	struct sigaction sa;
+	struct info I[NUM_EVENTS];
+	void * event_cookie[NUM_EVENTS];
+	int ret;
+	int i;
+	int done = 0;
+
+	WARNP_INIT;
+
+	(void)argc;
+	(void)argv;
+
+	/* Configure SIGUSR1 to cancel the loop. */
+	sa.sa_handler = events_interrupt;
+	sigemptyset(&sa.sa_mask);
+	sa.sa_flags = 0;
+	if (sigaction(SIGUSR1, &sa, NULL))
+		goto err0;
+
+	/* Queue some events. */
+	for (i = 0; i < NUM_EVENTS; i++) {
+		I[i].n = i;
+		event_cookie[i] = events_immediate_register(&event, &I[i], 10);
```

Comment:
I would register a single immediate event, and have that event register a new event.  That would effectively create an infinite loop since `events_run` would always decide that it has more immediate events to run.